# Data preprocessing - Creation of the final data frame for data analysis

## Introduction

This document contains the code for the item creation corresponding to the master thesis "...". 

The research question of the thesis is: Does the similarity of the first three noun phrase antecedents of double embedded sentences have an effect on the reading times of grammatical and ungrammatical double embedded sentences?

To examine this question the self-paced-reading (SPR) and eye-tracking data of Vasishth et al. (2010) are examined. The similarity of the three noun phrase antecedents is calculated with the machine learning model Word2vec for each sentence. A linear mixed model with varying intercepts for subjects is fitted that measures the effect of the interaction between the similarity and the grammaticality of the sentences.
In the following paragraphs the exact procedure for the creation of items as well as the actual code of item creation is presented. 

End product of this notebook: A data frame that contains the subject ID, the sentence number, the condition of the sentence, the reading time of the post V1 region (log transformed) and the corresponding similarity value (centred). The data was empirically collected for the Vasishth et al. (2010) study.

## Coding part

At first, the required packages have to be downloaded. Pandas and numpy are both libraries that provide functions for Python, enabling the user to deal with data or mathematical structures.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt # ist das nötig?

Next, the data corresponding to the first English SPR experiment of the Vasishth et al. (2010) study is uploaded into the notebook.

In [2]:
df = pd.read_csv('e1_en_spr_data.txt', sep=' ', names=['Subject_ID', 'Experiment', 'Sentence_Number', 'Condition', 'Word_Position','Word', 'RT'])
df

Subject_ID Experiment  Sentence_Number Condition  Word_Position  \
0               1   practice                1         -              0   
1               1   practice                1         -              1   
2               1   practice                1         -              2   
3               1   practice                1         -              3   
4               1   practice                1         -              4   
...           ...        ...              ...       ...            ...   
58741           9         E1               20         b             11   
58742           9         E1               20         b             12   
58743           9         E1               20         b             13   
58744           9         E1               20         b             14   
58745           9         E1               20         b             15   

             Word   RT  
0             The  602  
1           child  560  
2           liked  465  
3             the  434  
4      characters  425  
...           ...  ...  
58741      larger  194  
58742        than  178  
58743         the  194  
58744     average  194  
58745     spider.  218  

[58746 rows x 7 columns]

### Extraction of relevant data 

A lot of data points are not needed for the purpose of this study. Only the post V1 region of the 'gug' condition (the experimental condition) is needed. The relevant data has to be extracted of the data frame.

In [3]:
# Extract the experimental condition 'gug'
df_gug = df[df.Experiment=='gug']
df_gug

Subject_ID Experiment  Sentence_Number Condition  Word_Position  \
149             1        gug               14         b              0   
150             1        gug               14         b              1   
151             1        gug               14         b              2   
152             1        gug               14         b              3   
153             1        gug               14         b              4   
...           ...        ...              ...       ...            ...   
58725           9        gug                7         c              7   
58726           9        gug                7         c              8   
58727           9        gug                7         c              9   
58728           9        gug                7         c             10   
58729           9        gug                7         c             11   

             Word   RT  
149           The  354  
150       defence  330  
151           who  425  
152           the  386  
153         knife  402  
...           ...  ...  
58725    stranger  241  
58726         saw  322  
58727  questioned  258  
58728         the  265  
58729   customer.  266  

[10000 rows x 7 columns]

All words of the "gug" condition are now extracted and stored in a new data frame called df_gug. 
Since only the post V1 region is relevant for the data analysis the relevant rows, corresponding to the post V1 region, have to be extracted from the data frame df_gug. The post V1 region is always the second last word of every sentence.

How to get the second last word of every sentence:
Since all sentences differ in their length, it is not directly possible to access the second last word of each sentence. We therefore try first to find out the word positions of the last words of all sentences to be able to extract the second last words.

Our procedure to find the last word of every sentence works as follows: We start by finding the words that occur with a dot since every sentence ends with a dot. Therefore, the most straight forward way to find out where sentences end is to find out the positions of these dots. To do so, we create a Boolean series that gives us a TRUE when a dot occurs after a word and a FALSE when this is not the case.

In [4]:
last_word = df_gug.Word.apply(lambda x: '.' in x)
last_word

149      False
150      False
151      False
152      False
153      False
         ...  
58725    False
58726    False
58727    False
58728    False
58729     True
Name: Word, Length: 10000, dtype: bool

Next, we create an array 'index_last_word' that extracts the words' numbers that have a TRUE as output in our Boolean series (thus, that occur with a dot in the end). As a consequence, we have the indices of all last words of all sentences.

In [5]:
index_last_word = np.where(last_word)[0]

Since we don't need the last word but the second last word, we subtract 1 from each index and store the numbers in an array index_second_last. We then extract all words that correspond to the numbers in the array index_second_last.

In [6]:
# index of second last word
index_second_last = index_last_word -1

In [7]:
second_last_word = df_gug.iloc[index_second_last]
second_last_word

Subject_ID Experiment  Sentence_Number Condition  Word_Position Word  \
160             1        gug               14         b             11  the   
196             1        gug                6         b             11  the   
232             1        gug               10         b             11  the   
287             1        gug               15         c             10  the   
331             1        gug               13         a             11  the   
...           ...        ...              ...       ...            ...  ...   
58401           9        gug                8         d             10  the   
58430           9        gug                5         a             11  the   
58521           9        gug               13         a             11  the   
58692           9        gug                9         a             11  the   
58728           9        gug                7         c             10  the   

        RT  
160    473  
196    473  
232    617  
287    401  
331    682  
...    ...  
58401  321  
58430  234  
58521  250  
58692  218  
58728  265  

[800 rows x 7 columns]

### The similarity values

In this part we first import the nouns with two conditions, the similar and the contrast condition, and their corresponding similarity values. Next the concept of centred similarity is explained and carried out.

We now import the similarity values and store them into a data frame called Simval_df. 
We have two similarity values for each sentence. One similarity value is for the similar condition with three animate noun phrases, one for the contrast condition with two animate nouns and one inanimate noun in the middle. 
The similarity values are the mean similarity of the three nouns respectively.

In [8]:
Simval_df = pd.read_csv('df_SimilarityValues.csv', sep=',', names=['Sentence_Number', 'Similar_condition', 'Contrast_condition', 'SimVal_Similar', 'SimVal_Contrast'])
Simval_df

Sentence_Number                          Similar_condition  \
0                 0      ('carpenter', 'craftsman', 'peasant')   
1                 1           ('mother', 'daughter', 'sister')   
2                 2            ('worker', 'tenant', 'foreman')   
3                 3     ('trader', 'businessman', 'professor')   
4                 4          ('painter', 'musician', 'father')   
5                 5  ('saxophonist', 'trumpeter', 'conductor')   
6                 6     ('pharmacist', 'optician', 'stranger')   
7                 7           ('cleaner', 'janitor', 'doctor')   
8                 8          ('dancer', 'singer', 'bystander')   
9                 9           ('artist', 'sportsman', 'guard')   
10               10         ('clerk', 'bureaucrat', 'visitor')   
11               11               ('son', 'father', 'teacher')   
12               12     ('conductor', 'choirmaster', 'worker')   
13               13           ('defence', 'prosecutor', 'spy')   
14               14           ('cousin', 'brother', 'peasant')   
15               15          ('painter', 'musician', 'friend')   

                       Contrast_condition  SimVal_Similar  SimVal_Contrast  
0      ('carpenter', 'pillar', 'peasant')        0.391551         0.137571  
1             ('mother', 'gun', 'sister')        0.810291        -0.069852  
2         ('worker', 'bucket', 'foreman')        0.369354         0.225532  
3     ('trader', 'computer', 'professor')        0.623997         0.101565  
4            ('painter', 'hut', 'father')        0.423884         0.271721  
5   ('saxophonist', 'baton', 'conductor')        0.703727         0.308906  
6    ('pharmacist', 'button', 'stranger')        0.295159         0.233564  
7           ('cleaner', 'ball', 'doctor')        0.315141         0.207383  
8         ('dancer', 'shoe', 'bystander')        0.568988         0.452954  
9         ('artist', 'computer', 'guard')        0.253536         0.001793  
10           ('clerk', 'cane', 'visitor')        0.222973         0.076802  
11      ('son', 'loudspeaker', 'teacher')        0.700753         0.025134  
12      ('conductor', 'sponge', 'worker')        0.272818         0.304460  
13            ('defence', 'knife', 'spy')        0.457099         0.100111  
14       ('cousin', 'diamond', 'peasant')        0.521104        -0.052376  
15          ('painter', 'film', 'friend')        0.544250         0.194312

The concept of centred similarity is an equivalent to the sum contrast coding that will be used for the data analysis later on. 
In the sum contrast coding each data point is coded as a function of the overall mean of all data points.
For the sum contrast coding for the condition of grammaticality we have +1 for grammatical and -1 for ungrammatical sentences.
For the examination of the research question we have to carry out an interaction analysis to see the impact of the similarity values (hence the condition of similarity) on the condition of grammaticality. 
For this purpose, the similarity values have to be adapted to the sum contrast coding concept. This is done by subtracting the mean similarity value from each individual similarity value.

First we calculate the mean similarity value from the similar and the contrast condition.

In [9]:
Simval_Similar = list(Simval_df['SimVal_Similar'])
Simval_Contrast = list(Simval_df['SimVal_Contrast'])

In [10]:
added_sim = Simval_Similar + Simval_Contrast

In [11]:
mean_sim = np.mean(added_sim)
mean_sim

0.3123188643658068

Next, we need to subtract the mean from each similarity value.

This is done with a for loop that loops over the Simval_Similar list (containing all similarity values corresponding to the noun pairs of the similar group) or Simval_Contrast list (containing all similarity values corresponding to the noun pairs of the contrast group). For each value of the list, we subtract the mean from that value and store the result in a list called Centred_simval_similar or Centred_simval_contrast.

In [12]:
Centred_simval_similar = []
for value in Simval_Similar:
    Centred_simval_similar.append(value - mean_sim)
print(Centred_simval_similar)

[0.0792321980989073, 0.497972247103462, 0.05703535387874581, 0.3116784066951368, 0.1115649611747358, 0.39140784452320077, -0.017159967770567164, 0.0028219790256117094, 0.2566690714156721, -0.0587831987941172, -0.08934552324353714, 0.38843387077213265, -0.039501111110439524, 0.14478000710369088, 0.20878489772439934, 0.23193129608989693]


In [19]:
Centred_simval_contrast[13]

-0.2122081099951174

In [13]:
Centred_simval_contrast = []
for value in Simval_Contrast:
    Centred_simval_contrast.append(value - mean_sim)
print(Centred_simval_contrast)

[-0.1747479225450661, -0.38217081417678855, -0.0867871894442942, -0.21075361417024396, -0.04059784402488731, -0.003413009544601664, -0.07875486006378196, -0.10493575324653648, 0.140635010699043, -0.3105253755056765, -0.23551733707427047, -0.2871846380585339, -0.007859083911171183, -0.2122081099951174, -0.36469443430542015, -0.11800735731958412]


For our end data frame we need log transformed reading times.

In [14]:
logRT = list(np.log(second_last_word['RT']))

The next section is about merging the correct similarity value to its corresponding sentence number,  condition and reading time. 

To do that, we first create a data frame containing all the relevant information for our end data frame, except of the similarity values and the contrast coding for the condition of grammaticality (+1 and -1).

In [15]:
df_WithoutSim = pd.DataFrame(list(zip(second_last_word.Subject_ID, second_last_word.Sentence_Number, second_last_word.Condition, logRT)),
                            columns = ['Subject_ID', 'Sentence_Number', 'Condition', 'logRT'])
df_WithoutSim

Subject_ID  Sentence_Number Condition     logRT
0             1               14         b  6.159095
1             1                6         b  6.159095
2             1               10         b  6.424869
3             1               15         c  5.993961
4             1               13         a  6.525030
..          ...              ...       ...       ...
795           9                8         d  5.771441
796           9                5         a  5.455321
797           9               13         a  5.521461
798           9                9         a  5.384495
799           9                7         c  5.579730

[800 rows x 4 columns]

The data frame we need for the statistical analysis would have one more column containing the centred similarity values. We have two similarity values for all sentences (1-16). Condition a and c need the similarity values of the similar condition, condition b and d need the similarity value of the contrast condition.

To do that, we write a for loop that loops over the list “Sentence_Number” of our data frame “df_WithoutSim”. Note that the list “Sentence_Number” does not only contain the numbers from 1 to 16! It has length 800 and contains the sentence numbers that correspond to the sentences that have been read by a specific participant. The ‘iterater’ variable starts to count from zero until the end of a list. In other words, it assigns an index, stating at zero until the end, for every time we move on to the next item of our for loop. The variable ‘i’ refers to the entry at a specific point in the list “Sentence_Number”. The function enumerate makes the for loop loop through the whole list “Sentence_Number”.  Assume we start at the first entry in Sentence_Number. The first entry is the sentence 14, read by participant 1. Thus, our iterater would start counting at zero and the variable ‘i’ would get 14. Before jumping to the next entry, we now want to know which condition this entry has. This is important in order to assign the correct similarity value to that entry. Therefore, we look at the entry in the “Condition” column of our data frame “df_WithoutSim” in the same row (in our example row zero). If the entry is either ‘a’ or ‘c’ (similar condition) we append the centred similarity value for the similar group of that same sentence in an empty list called “Similarity_Values_dat”. In our example we would append the centred similarity value for the contrast condition of sentence 14 ~-0.212, since the condition in the first entry of the data frame is 'b'. Thereby, we have to take the entry i-1 of the “Centred_simval_similar” list. We have to take i-1 since Python starts counting at zero. ‘i’ is the correct sentence number. In our example i=14. In the “Centred_simval_similar” list we have all the similarity values for all sentences in the correct order, starting from zero to fifteen. If we want to add the similarity value of the fourteenth sentence, we have to take the thirteenth entry of the “Centred_simval_similar” list. If the condition is not ‘a’ or ‘c’ (meaning it has to be ‘b’ or ‘d’) we want to add the corresponding contrast centred similarity value. We then take the i-1th entry of the “Centered_simval_contrast” list and append it to the “Similarity_Values_dat” list. After that, enumerate makes us loop to the next entry of “Sentence_Number” etc.

In [18]:
Similarity_Values_dat = []

for iterater, i in enumerate(df_WithoutSim.Sentence_Number):
    if df_WithoutSim.Condition[iterater] in ['a', 'c']:
        simVal_similar = Centred_simval_similar[i-1]
        Similarity_Values_dat.append(simVal_similar)
    else:
        simVal_Con = Centred_simval_contrast[i-1]
        Similarity_Values_dat.append(simVal_Con)
print(Similarity_Values_dat)

[-0.2122081099951174, -0.003413009544601664, -0.3105253755056765, 0.20878489772439934, -0.039501111110439524, 0.1115649611747358, -0.017159967770567164, 0.05703535387874581, 0.2566690714156721, -0.2871846380585339, -0.38217081417678855, -0.10493575324653648, -0.21075361417024396, -0.11800735731958412, -0.08934552324353714, 0.0792321980989073, 0.0028219790256117094, -0.0867871894442942, 0.3116784066951368, 0.23193129608989693, 0.38843387077213265, -0.007859083911171183, 0.497972247103462, -0.04059784402488731, -0.23551733707427047, -0.1747479225450661, -0.36469443430542015, -0.0587831987941172, 0.14478000710369088, 0.39140784452320077, 0.140635010699043, -0.07875486006378196, 0.20878489772439934, -0.003413009544601664, 0.2566690714156721, -0.38217081417678855, -0.21075361417024396, -0.3105253755056765, -0.08934552324353714, -0.11800735731958412, -0.2871846380585339, 0.1115649611747358, -0.2122081099951174, 0.05703535387874581, 0.0792321980989073, -0.10493575324653648, -0.017159967770567

For the final data frame for data analysis we need the grammaticality condition to be coded as + and - 1. Conditions a and b refer to grammatical sentences, condition c and d to ungrammatical sentences. We iterate over the column "Condition" of our data frame 'df_WithoutSim'. If the condition is equal to 'a' or 'b' we enter a 1 in the list 'Contrast_Coding', else (if condition is equal to 'c' or 'd') we enter -1 in the list. This list can later be added in the data frame in a new column called "Contrast_coding".

In [20]:
Contrast_Coding = []

for cond in df_WithoutSim.Condition:
    if cond in ['a','b']:
        Contrast_Coding.append(1)
    else:
        Contrast_Coding.append(-1)
print(Contrast_Coding)

[1, 1, 1, -1, 1, 1, -1, -1, 1, -1, 1, -1, -1, -1, -1, 1, 1, -1, 1, 1, 1, 1, -1, 1, -1, 1, -1, -1, -1, -1, 1, -1, 1, -1, -1, -1, 1, -1, 1, 1, 1, -1, -1, 1, -1, 1, 1, -1, -1, -1, 1, 1, 1, -1, 1, -1, -1, -1, -1, 1, -1, 1, 1, 1, -1, 1, 1, 1, -1, -1, -1, 1, 1, -1, -1, 1, 1, -1, 1, -1, -1, -1, 1, 1, 1, -1, 1, -1, 1, 1, -1, -1, -1, -1, 1, 1, -1, 1, 1, -1, -1, -1, 1, -1, 1, 1, 1, 1, -1, 1, -1, -1, -1, 1, 1, -1, 1, 1, -1, -1, 1, -1, 1, -1, 1, 1, -1, -1, -1, -1, 1, 1, 1, 1, 1, -1, -1, -1, -1, -1, 1, -1, 1, 1, -1, 1, -1, 1, -1, 1, -1, -1, -1, 1, 1, 1, 1, -1, 1, -1, -1, 1, 1, 1, 1, -1, 1, -1, 1, -1, -1, 1, 1, -1, -1, -1, -1, -1, 1, -1, 1, 1, -1, -1, -1, 1, 1, 1, -1, 1, 1, -1, 1, 1, 1, -1, -1, -1, 1, -1, -1, 1, -1, 1, -1, -1, 1, 1, 1, 1, 1, -1, 1, 1, -1, 1, -1, -1, -1, -1, -1, -1, 1, 1, 1, 1, -1, -1, -1, 1, 1, -1, -1, 1, 1, -1, 1, -1, -1, 1, 1, 1, 1, -1, 1, -1, -1, 1, 1, -1, -1, 1, -1, -1, -1, 1, 1, 1, -1, -1, -1, -1, -1, -1, -1, 1, 1, 1, 1, -1, 1, 1, 1, -1, -1, 1, 1, 1, -1, -1, -1, 1, -1, -1, -1, 

Now we have everything we need to create the final data frame, thus to add a column for the contrast coding numbers and the centred similarity values.

In [21]:
df_WithSim = pd.DataFrame(list(zip(df_WithoutSim.Subject_ID, df_WithoutSim.Sentence_Number, df_WithoutSim.Condition, Contrast_Coding, logRT, Similarity_Values_dat)),
                            columns = ['Subject_ID', 'Sentence_Number', 'Condition', 'Contrast_Coding','logRT', 'Centered_Sim_Val'])
df_WithSim

Subject_ID  Sentence_Number Condition  Contrast_Coding     logRT  \
0             1               14         b                1  6.159095   
1             1                6         b                1  6.159095   
2             1               10         b                1  6.424869   
3             1               15         c               -1  5.993961   
4             1               13         a                1  6.525030   
..          ...              ...       ...              ...       ...   
795           9                8         d               -1  5.771441   
796           9                5         a                1  5.455321   
797           9               13         a                1  5.521461   
798           9                9         a                1  5.384495   
799           9                7         c               -1  5.579730   

     Centered_Sim_Val  
0           -0.212208  
1           -0.003413  
2           -0.310525  
3            0.208785  
4           -0.039501  
..                ...  
795         -0.104936  
796          0.111565  
797         -0.039501  
798          0.256669  
799         -0.017160  

[800 rows x 6 columns]

In [20]:
# We are now ready to export the data frame for the data analysis in r
df_WithSim.to_csv('Final_DataFrame_e1.csv', sep=',', header=False, index=True)